From fb05106e99511b3d72e8ba41fe29b1b28f83864e Mon Sep 17 00:00:00 2001
From: ZhangPeng <zhangpeng362@huawei.com>
Date: Tue, 23 May 2023 19:12:17 +0800
Subject: [PATCH 046/263] [Huawei] userswap: fix kernel BUG in
 userfaultfd_release()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I6CAIM

--------------------------------

During the unregister process, the vm_flags of userfaultfd
(VM_UFFD_MISSING | VM_UFFD_WP) are cleaned up, and
vma->vm_userfaultfd_ctx.ctx is set to NULL_VM_UFFD_CTX. However,
VM_USWAP is not cleared. If ctx is set to NULL_VM_UFFD_CTX and VM_USWAP
is not cleared, a BUG_ON() will occur in userfaultfd_release().
To fix this bug, clear the VM_USWAP flag in userfaultfd_unregister().

Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
---
 fs/userfaultfd.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/fs/userfaultfd.c b/fs/userfaultfd.c
index bee6067b9c52..a4977ef8084f 100644
--- a/fs/userfaultfd.c
+++ b/fs/userfaultfd.c
@@ -1507,7 +1507,7 @@ static int userfaultfd_unregister(struct userfaultfd_ctx *ctx,
 	struct vm_area_struct *vma, *prev, *cur;
 	int ret;
 	struct uffdio_range uffdio_unregister;
-	unsigned long new_flags;
+	unsigned long new_flags, userfault_flags;
 	bool found;
 	unsigned long start, end, vma_end;
 	const void __user *buf = (void __user *)arg;
@@ -1608,8 +1608,11 @@ static int userfaultfd_unregister(struct userfaultfd_ctx *ctx,
 			range.len = vma_end - start;
 			wake_userfault(vma->vm_userfaultfd_ctx.ctx, &range);
 		}
-
-		new_flags = vma->vm_flags & ~(VM_UFFD_MISSING | VM_UFFD_WP);
+		userfault_flags = VM_UFFD_MISSING | VM_UFFD_WP;
+#ifdef CONFIG_USERSWAP
+		uswap_release(&userfault_flags);
+#endif
+		new_flags = vma->vm_flags & ~userfault_flags;
 		prev = vma_merge(mm, prev, start, vma_end, new_flags,
 				 vma->anon_vma, vma->vm_file, vma->vm_pgoff,
 				 vma_policy(vma),
-- 
2.17.1

